rfcs/0003-timestamps-and-ttl.md

```diff
@@ -0,0 +1,363 @@
+   keys at different timestamps to allow for similar "historical" read operations.
+
+## Goals
+
+- Define the API for specifying timestamps via user defined clocks
+- Define the changes to the on-disk format of stored rows
+- Outline mechanism to support Time-To-Live (TTL) via Compaction Filters
+
+## Out of Scope
+
+While the following are out of scope for the initial implementation, the design should
+not preclude:
+
+- Handling out-of-order insertions is not in scope for this RFC, though we make sure the design
+  accounts for the future possibility (see rejected alternatives section)
+- Outlining compaction strategies that are optimized for TTL (such as Cassandra's
+  [TWCS](https://cassandra.apache.org/doc/stable/cassandra/operating/compaction/twcs.html)). This
+  will require additional metadata per-SST and a compaction strategy to match, but is otherwise
+  possible with what's outlined in the scope of this design.
+- Outlining compaction strategies that target efficient deletes (such as outlined
+  in [Lethe](https://disc-projects.bu.edu/lethe/))
+- Ensuring deletions are resolved within a time bound (a TTL does not ensure the data is deleted
+  from the object store within the specified time)
+- A public API for compaction filters (see https://github.com/slatedb/slatedb/issues/225) - this RFC
+  will implement TTL using a compaction filter but will not yet expose a public API for plugging in
+  custom compaction filters. Instead, the implementation should ensure that such an extension is
+  easy.
+
+## Public API
+
+We will make the following changes to `DbOptions`:
+
+- introduce a configuration for a custom clock, and will default to using the system clock
+- introduce a configuration for a default TTL, which defaults to `None`
+
+Note that the clock is expected to return monotonically increasing (greater than or equal to
+previous) values. This will be enforced by wrapping the supplied `Clock` implementation that
+tracks the previous values returned.
+
+```rust
+/// defines the clock that SlateDB will use during this session
+pub trait Clock {
+    /// Returns a timestamp in milliseconds since the unix epoch,
+    /// must return monotonically increasing numbers (this is enforced
+    /// at runtime and will panic if the invariant is broken)
+    ///
+    /// Note that this clock does not need to return a number that
+    /// represents the unix timestamp; the only requirement is that
+    /// it represents a sequence that can attribute a logical ordering
+    /// to actions on the database
+    fn now(&self) -> i64;
+}
+
+/// contains the default implementation of the Clock, and will return the system time
+pub struct SystemClock;
+
+// ...
+
+pub struct DbOptions {
+    // ...
+
+    /// The Clock to use for insertion timestamps
+    ///
+    /// Default: the default clock uses the local system time on the machine
+    clock: Option<Arc<dyn Clock>>,
```

Comment:
could i understand this Clock as a `Wallclock` which gives the wall time in unix timestamp, not a monotonic logical clock (which might be an increasing atomic integer), right?

Reply:
Yes. For now this RFC suggests enforcing monotonic increments on the tick (see the comment in the `Clock` interface above):

```rust
    /// Returns a timestamp in milliseconds since the unix epoch,
    /// must return monotonically increasing numbers (this is enforced
    /// at runtime and will panic if the invariant is broken)
    ///
    /// Note that this clock does not need to return a number that
    /// represents the unix timestamp; the only requirement is that
    /// it represents a sequence that can attribute a logical ordering
    /// to actions on the database
```

Eventually the idea is to lift this requirement when we figure out how we want to handle out-of-order insertions.

We will probably need a separate logical clock for transactions.